docs/PROTOCOL.md

```diff
@@ -735,70 +740,81 @@ uint24 | Device token | 3 | Token that has been registered.
+uint8 | Protocol | 1 | Protocol of the microapp command and result packets, should match what you get from the [microapp info packet](#microapp-info-packet).
+uint8 | App index | 1 | Index of the microapp on this Crownstone.
 
-The packets are all of the same size. Important! The data in the last packet has to contain `0xFF` for the values beyond the application size. This is due to the fact that on flash writing `0xFF` does not change the flash. 
+#### Microapp upload packet
 
-The total maximum size of a binary (256 packets with each 49 bytes) is around 12 kB. This is considered sufficient
-for this first version. Larger binaries would require more stringent containerization as well to make sure its not
-eating too many other resources.
-
-#### Validate microapp
-
-![Send microapp validate packet](../docs/diagrams/send_microapp_validate_packet.png)
+![Microapp upload packet](../docs/diagrams/microapp_upload_packet.png)
 
 Type | Name | Length | Description
 --- | --- | --- | ---
-uint8 | Protocol | 1 | Protocol version of binary sends (default 0).
-uint8 | App ID | 1 | Application identifier. In theory, multiple apps can be supported (not supported yet).
-uint8 | Opcode | 1 | `CS_MICROAPP_OPCODE_VALIDATE` (`0x02`).
-uint16 | Size | 2 | Size in bytes of the complete binary.
-uint16 | Checksum | 2 | The checksum of the complete program.
+[Microapp header](#microapp-header-packet) | Header | 2 |
+uint16 | Offset | 2 | Offset in bytes of this chunk of data. Must be a multiple of 4.
+uint8[] | Data chunk | N | A chunk of the binary.
+
+For example, if the microapp binary is 299 byte, and your chunks are 128 byte, then you have 3 upload commands with offsets: 0, 128, 256. The last command will have 44 byte of data: the remaining 43 byte, padded to 44 byte, so it is a multiple of 4.
 
-After sending the last packet, the microapp has to be validated.
+From the get info command, you know which index to use, and what protocol, sdk, and sizes are supported.
 
-#### Enable/disable microapp
+When result is ERR_WAIT_FOR_SUCCESS, you have to wait for the result to change to ERR_SUCCESS.
+In case the stored data matches what you upload, you will get ERR_SUCCESS_NO_CHANGE.
+If the stored data does not match, you will get ERR_WRITE_DISABLED, meaning you should first remove the current microapp.
 
-![Send microapp enable packet](../docs/diagrams/send_microapp_enable_packet.png)
+#### Microapp info packet
+
+![Microapp info packet](../docs/diagrams/microapp_info_packet.png)
 
 Type | Name | Length | Description
 --- | --- | --- | ---
-uint8 | Protocol | 1 | Protocol version of binary sends (default 0).
-uint8 | App ID | 1 | Application identifier. In theory, multiple apps can be supported (not supported yet).
-uint8 | Opcode | 1 | `CS_MICROAPP_OPCODE_ENABLE` (`0x03`) or `CS_MICROAPP_OPCODE_DISABLE` (`0x04`).
-uint16 | Offset | 2 | Offset of the `dummy_main` function into the binary.
+uint8 | Protocol | 1 | Supported microapp protocol.
```

Comment:
Can we be clearer what it means if different versions are returned here?
```suggestion
uint8 | Protocol | 1 | current protocol version of bluenet code firmware interpreting the microapp.
```

Reply:
Changed to "Supported protocol of the microapp command and result packets". Like the one in the header packet.